engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/distribute/DeploymentRedistributor.java

```diff
@@ -35,15 +48,57 @@ public void onRecovered(final ReadonlyStreamProcessorContext context) {
     if (context.getPartitionId() != DEPLOYMENT_PARTITION) {
       return;
     }
-    final var deploymentDistributionBehavior =
+    deploymentDistributionBehavior =
         new DeploymentDistributionBehavior(
             context.getWriters(), partitionsCount, deploymentDistributionCommandSender);
 
-    final var schedulingService = context.getScheduleService();
-    schedulingService.runAtFixedRate(
-        DEPLOYMENT_REDISTRIBUTION_INTERVAL,
-        () ->
-            deploymentState.foreachPendingDeploymentDistribution(
-                deploymentDistributionBehavior::distributeDeploymentToPartition));
+    context
+        .getScheduleService()
+        .runAtFixedRate(DEPLOYMENT_REDISTRIBUTION_INTERVAL, this::runRetryCycle);
   }
+
+  private void runRetryCycle() {
+    final var pendingDistributions = new HashSet<PendingDistribution>();
+    deploymentState.foreachPendingDeploymentDistribution(
+        (deploymentKey, partitionId, directBuffer) -> {
+          final var pending = new PendingDistribution(deploymentKey, partitionId);
+          pendingDistributions.add(pending);
+          retryDistribution(pending, directBuffer);
+        });
+    // Remove attempts for completed distributions
+    distributionAttempts.keySet().removeIf(Predicate.not(pendingDistributions::contains));
+  }
+
+  private void retryDistribution(
+      final PendingDistribution pending, final DirectBuffer copiedDeploymentBuffer) {
+    if (!shouldRetryNow(pending)) {
+      return;
+    }
+
+    LOG.info(
+        "Retrying to distribute deployment {} to partition {}",
+        pending.deploymentKey,
+        pending.partitionId);
+    deploymentDistributionBehavior.distributeDeploymentToPartition(
+        pending.deploymentKey, pending.partitionId, copiedDeploymentBuffer);
+  }
+
+  private boolean shouldRetryNow(final PendingDistribution distributionKey) {
+    // attempt starts off at 0, ensuring that we wait between DEPLOYMENT_REDISTRIBUTION_INTERVAL and
+    // 2 * DEPLOYMENT_REDISTRIBUTION_INTERVAL before retrying distribution.
+    final long attempt =
+        distributionAttempts.compute(
+            distributionKey, (k, attempts) -> attempts != null ? attempts + 1 : 0L);
+
+    if (attempt >= RETRY_MAX_BACKOFF_ATTEMPTS) {
+      // Retry in intervals of RETRY_MAX_BACKOFF_DURATION
+      return attempt % RETRY_MAX_BACKOFF_ATTEMPTS == 0;
+    } else {
+      // Retry in intervals of DEPLOYMENT_REDISTRIBUTION_INTERVAL
+      // The interval is doubling until we reached RETRY_MAX_BACKOFF_DURATION
+      return Long.bitCount(attempt) == 1;
```

Comment:
😄 I would've never thought of this.

Reply:
:smile: I love that trick, so much nicer than fiddling with logarithms and such.